# Dataset

## Download dataset
Vietnamese Students' Feedback Corpus (UIT-VSFC) is the resource consists of over 16,000 sentences which are human-annotated with two different tasks: sentiment-based and topic-based classifications.

[1] Kiet Van Nguyen, Vu Duc Nguyen, Phu Xuan-Vinh Nguyen, Tham Thi-Hong Truong, Ngan Luu-Thuy Nguyen, UIT-VSFC: Vietnamese Students' Feedback Corpus for Sentiment Analysis,  2018 10th International Conference on Knowledge and Systems Engineering (KSE 2018), November 1-3, 2018, Ho Chi Minh City, Vietnam

In [2]:
!pip install datasets

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 491.2/491.2 kB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 5.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 183.9/183.9 kB 9.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 5.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 8.8 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2025.3.0
    Uninstalling fsspec-2025.3.0:
      Successfully uninstalled fsspec-2025.3.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gcsfs 2025.3.0 requires fsspec==2025.3.0, but you have fsspec 2024.12.0 which is incompatible.
torch 2.6.0+cu124 requires nvidia-cublas-cu12==12.4.5.8; platform_system == "Linux" and platform_machine == "x86_64", but you have nvidia-cublas-cu12 12.5.3.2 which is i

In [3]:
from datasets import load_dataset

dataset = load_dataset("uitnlp/vietnamese_students_feedback")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/4.79k [00:00<?, ?B/s]

vietnamese_students_feedback.py:   0%|          | 0.00/4.97k [00:00<?, ?B/s]

0000.parquet:   0%|          | 0.00/475k [00:00<?, ?B/s]

0000.parquet:   0%|          | 0.00/63.3k [00:00<?, ?B/s]

0000.parquet:   0%|          | 0.00/134k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/11426 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/1583 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/3166 [00:00<?, ? examples/s]

## Interacting with the downloaded data

In [4]:
dataset['train'][:5]

{'sentence': ['slide giáo trình đầy đủ .',
  'nhiệt tình giảng dạy , gần gũi với sinh viên .',
  'đi học đầy đủ full điểm chuyên cần .',
  'chưa áp dụng công nghệ thông tin và các thiết bị hỗ trợ cho việc giảng dạy .',
  'thầy giảng bài hay , có nhiều bài tập ví dụ ngay trên lớp .'],
 'sentiment': [2, 2, 0, 0, 2],
 'topic': [1, 0, 1, 0, 0]}

In [5]:
train_set = dataset['train']
train_set

Dataset({
    features: ['sentence', 'sentiment', 'topic'],
    num_rows: 11426
})

In [ ]:
train_set[0]

In [6]:
len(train_set)

11426

## Split a sentence

In [7]:
# Read a sentence
example_word_list = train_set[0]['sentence']
example_word_list

'slide giáo trình đầy đủ .'

In [8]:
# Split sentence word-by-word
example_word_list.split()

['slide', 'giáo', 'trình', 'đầy', 'đủ', '.']

In [9]:
# Join words into 1 full sentence
sentence = ""
for word in example_word_list:
    sentence += word
sentence

'slide giáo trình đầy đủ .'

In [10]:
# Get 10 sentences to process
sentence_list = []
for idx in range(10):
    sentence = ""
    for word in train_set[idx]['sentence']:
        sentence += word
    sentence_list.append(sentence)
sentence_list

['slide giáo trình đầy đủ .',
 'nhiệt tình giảng dạy , gần gũi với sinh viên .',
 'đi học đầy đủ full điểm chuyên cần .',
 'chưa áp dụng công nghệ thông tin và các thiết bị hỗ trợ cho việc giảng dạy .',
 'thầy giảng bài hay , có nhiều bài tập ví dụ ngay trên lớp .',
 'giảng viên đảm bảo thời gian lên lớp , tích cực trả lời câu hỏi của sinh viên , thường xuyên đặt câu hỏi cho sinh viên .',
 'em sẽ nợ môn này , nhưng em sẽ học lại ở các học kỳ kế tiếp .',
 'thời lượng học quá dài , không đảm bảo tiếp thu hiệu quả .',
 'nội dung môn học có phần thiếu trọng tâm , hầu như là chung chung , khái quát khiến sinh viên rất khó nắm được nội dung môn học .',
 'cần nói rõ hơn bằng cách trình bày lên bảng thay vì nhìn vào slide .']

# Text processing

## N-grams
- N-grams are continuous sequences of words or symbols, or tokens in a document. In technical terms, they can be defined as the neighboring sequences of items in a document.
- We can use n-grams or multiple other text preprocessing algorithms by incorporating [`nltk`](https://www.nltk.org/) library.

In [11]:
example_sentence = [
    'This is the first document.',
    'This document is the second document.',
    'And this is the third one.',
    'Is this the first document?',
]

In [12]:
from nltk import ngrams
import numpy as np

num_of_grams = np.arange(1, 4, 1) # Test 3 n-grams

print("Original sentence:", example_sentence[1])
print("==="*5)

for gram in num_of_grams:
    gram = int(gram)
    splitted_sentence = ngrams(example_sentence[1].split(), gram)
    print(f"{gram}-gram: ",end ='')
    n_grams_list = [ ' '.join(grams) for grams in splitted_sentence]
    print(n_grams_list)
    print()

Original sentence: This document is the second document.
1-gram: ['This', 'document', 'is', 'the', 'second', 'document.']

2-gram: ['This document', 'document is', 'is the', 'the second', 'second document.']

3-gram: ['This document is', 'document is the', 'is the second', 'the second document.']



In [13]:
splitted_sentence = ngrams(train_set[0]['sentence'].split(), 2)
for grams in splitted_sentence:
    print(grams)

('slide', 'giáo')
('giáo', 'trình')
('trình', 'đầy')
('đầy', 'đủ')
('đủ', '.')


## Extract features with n-grams

In [14]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

In [15]:
count_vectorize_model = CountVectorizer(ngram_range = (1, 1))
n_grams_feature_vector = count_vectorize_model.fit_transform(example_sentence).toarray()
word_frequency = pd.DataFrame(data = n_grams_feature_vector, columns = count_vectorize_model.get_feature_names_out())
word_frequency.T

0  1  2  3
and       0  0  1  0
document  1  2  0  1
first     1  0  0  1
is        1  1  1  1
one       0  0  1  0
second    0  1  0  0
the       1  1  1  1
third     0  0  1  0
this      1  1  1  1

In [16]:
count_vectorize_model = CountVectorizer(ngram_range = (1, 1))

n_grams_feature_vector = count_vectorize_model.fit_transform([sentence_list[5]]).toarray()

word_frequency = pd.DataFrame(data = n_grams_feature_vector, columns = count_vectorize_model.get_feature_names_out())

print('Example sentence:', sentence_list[5])
word_frequency

Example sentence: giảng viên đảm bảo thời gian lên lớp , tích cực trả lời câu hỏi của sinh viên , thường xuyên đặt câu hỏi cho sinh viên .


bảo  cho  câu  của  cực  gian  giảng  hỏi  lên  lớp  lời  sinh  thường  \
0    1    1    2    1    1     1      1    2    1    1    1     2       1   

   thời  trả  tích  viên  xuyên  đảm  đặt  
0     1    1     1     3      1    1    1

In [17]:
count_vectorize_model = CountVectorizer(ngram_range = (1, 2))

n_grams_feature_vector = count_vectorize_model.fit_transform(example_sentence).toarray()

word_frequency = pd.DataFrame(data = n_grams_feature_vector, columns = count_vectorize_model.get_feature_names_out())
word_frequency.T

0  1  2  3
and              0  0  1  0
and this         0  0  1  0
document         1  2  0  1
document is      0  1  0  0
first            1  0  0  1
first document   1  0  0  1
is               1  1  1  1
is the           1  1  1  0
is this          0  0  0  1
one              0  0  1  0
second           0  1  0  0
second document  0  1  0  0
the              1  1  1  1
the first        1  0  0  1
the second       0  1  0  0
the third        0  0  1  0
third            0  0  1  0
third one        0  0  1  0
this             1  1  1  1
this document    0  1  0  0
this is          1  0  1  0
this the         0  0  0  1

## Problem set 1
Based on the UIT-VSFC dataset and the aforementioned information.
- Create an $n$-gram word frequency table, such that $n$ could be any number of your desire.
- With $n=1$ and $n=2$, what is the most popular word in the dataset ?
- With $n=1$ and $n=2$, what is the rarest word in the dataset ?
- What are the limitations of this data processing flow ? How can we overcome those ?


### Retrieve all sentences within the dataset

In [18]:
from typing import List

def get_all_sentences(dataset) -> List[str]:
    """
    Function to get all sentences and store them into a list of strings

    Args:
    dataset -- The subset (i.e., train/valid/test) in UIT-VSFC dataset

    Returns:
    A list of all sentences in a subset data of the UIT-VSFC.
    """

    list_all_sentence: list = []

    for idx in range(len(dataset)):
        sentence = ""
        for word in dataset[idx]['sentence']:
            sentence += word
        list_all_sentence.append(sentence)


    return list_all_sentence

In [19]:
list_all_sentence: list = get_all_sentences(train_set)
print(f"#sentences within the dataset: {len(list_all_sentence)}")
print(f"Example sentence: {list_all_sentence[:5]}")

#sentences within the dataset: 11426
Example sentence: ['slide giáo trình đầy đủ .', 'nhiệt tình giảng dạy , gần gũi với sinh viên .', 'đi học đầy đủ full điểm chuyên cần .', 'chưa áp dụng công nghệ thông tin và các thiết bị hỗ trợ cho việc giảng dạy .', 'thầy giảng bài hay , có nhiều bài tập ví dụ ngay trên lớp .']


### Build the word frequency table

In [20]:
def n_gram_word_frequency(sentence_list: list,
                          n: int) -> pd.DataFrame:
    """
    Function to build a word frequency table based on n-grams

    Args:
    sentence_list (list) -- A list of all sentences needed for table constructing process
    n (int) -- Number of grams that we parse into this function

    Returns:
    A dataframe contains all words after conducting n-grams and their respective frequencies
    """

    ### YOUR CODE STARTS HERE

    count_vectorize_model = CountVectorizer(ngram_range = (n, n))
    n_grams_feature_vector = count_vectorize_model.fit_transform(sentence_list).toarray()
    word_frequency_table = pd.DataFrame(data = n_grams_feature_vector, columns = count_vectorize_model.get_feature_names_out())

    ### YOUR CODE ENDS HERE

    return word_frequency_table

#### **Với n = 1**

In [21]:
# Construct the table of word frequency
word_frequency_table_unigram = n_gram_word_frequency(sentence_list=list_all_sentence, n=1)
word_frequency_table_unigram

10  100  10h  10h30  11  11doubledot55  11h30  11h55  12  \
0       0    0    0      0   0              0      0      0   0   
1       0    0    0      0   0              0      0      0   0   
2       0    0    0      0   0              0      0      0   0   
3       0    0    0      0   0              0      0      0   0   
4       0    0    0      0   0              0      0      0   0   
...    ..  ...  ...    ...  ..            ...    ...    ...  ..   
11421   0    0    0      0   0              0      0      0   0   
11422   0    0    0      0   0              0      0      0   0   
11423   0    0    0      0   0              0      0      0   0   
11424   0    0    0      0   0              0      0      0   0   
11425   0    0    0      0   0              0      0      0   0   

       12doubledot00  ...  ấy  ẩn  ắt  ốc  ồn  ổn  ủa  ủng  ức  ứng  
0                  0  ...   0   0   0   0   0   0   0    0   0    0  
1                  0  ...   0   0   0   0   0   0   0    0   0    0  
2                  0  ...   0   0   0   0   0   0   0    0   0    0  
3                  0  ...   0   0   0   0   0   0   0    0   0    0  
4                  0  ...   0   0   0   0   0   0   0    0   0    0  
...              ...  ...  ..  ..  ..  ..  ..  ..  ..  ...  ..  ...  
11421              0  ...   0   0   0   0   0   0   0    0   0    0  
11422              0  ...   0   0   0   0   0   0   0    0   0    0  
11423              0  ...   0   0   0   0   0   0   0    0   0    0  
11424              0  ...   0   0   0   0   0   0   0    0   0    0  
11425              0  ...   0   0   0   0   0   0   0    0   0    0  

[11426 rows x 2459 columns]

 - Ta kiểm tra bằng cách in ra dòng đầu tiên của bảng tần số, tương ứng với câu đầu tiên trong bộ dữ liệu train.

In [22]:
word_frequency_table_unigram.iloc[0].sort_values(ascending=False)

slide     1
đầy       1
trình     1
đủ        1
giáo      1
         ..
khoát     0
khoăn     0
khoảng    0
khu       0
khiến     0
Name: 0, Length: 2459, dtype: int64

- Ta sẽ in ra toàn bộ bảng tần suất. Sau đó, ta sẽ các giá trị tần số theo cột và sắp xếp để tìm ra các từ có tần số xuấn hiện nhiều nhất và ít nhất

In [23]:
# Những từ ít xuất hiện nhất
word_frequency_sorted_unigram = word_frequency_table_unigram.sum(axis=0).sort_values()
print(word_frequency_sorted_unigram[:30])

upload    1
tẻ        1
tết       1
hỏa       1
hố        1
hạ        1
hạng      1
hạt       1
unit      1
tẳng      1
user      1
uy        1
tồi       1
tồn       1
tội       1
tớ        1
tờ        1
hầm       1
tủ        1
họng      1
hỉnh      1
hẹn       1
hứa       1
hở        1
tạt       1
tạm       1
tố        1
tỏ        1
tôt       1
hỗn       1
dtype: int64


In [24]:
# Những từ xuất hiện nhiều nhất
print(word_frequency_sorted_unigram[-10:])

và       2068
không    2177
tình     2266
bài      2336
học      2940
sinh     3082
thầy     3095
dạy      3156
giảng    3711
viên     4803
dtype: int64


**Kết luận**:
-  Với n = 1, từ xuất hiện nhiều nhất là "viên", những từ xuất hiện ít nhất như: "upload", "tẻ", "tết",...

#### **Với n = 2**

In [25]:
word_frequency_table_bigram = n_gram_word_frequency(sentence_list=list_all_sentence, n=2)
word_frequency_table_bigram


10 50  10 bài  10 fraction  10 kiến  10 luôn  10 mấy  10 mới  10 người  \
0          0       0            0        0        0       0       0         0   
1          0       0            0        0        0       0       0         0   
2          0       0            0        0        0       0       0         0   
3          0       0            0        0        0       0       0         0   
4          0       0            0        0        0       0       0         0   
...      ...     ...          ...      ...      ...     ...     ...       ...   
11421      0       0            0        0        0       0       0         0   
11422      0       0            0        0        0       0       0         0   
11423      0       0            0        0        0       0       0         0   
11424      0       0            0        0        0       0       0         0   
11425      0       0            0        0        0       0       0         0   

       10 năm  10 phút  ...  ứng kịp  ứng nhu  ứng nhưng  ứng tốt  ứng yêu  \
0           0        0  ...        0        0          0        0        0   
1           0        0  ...        0        0          0        0        0   
2           0        0  ...        0        0          0        0        0   
3           0        0  ...        0        0          0        0        0   
4           0        0  ...        0        0          0        0        0   
...       ...      ...  ...      ...      ...        ...      ...      ...   
11421       0        0  ...        0        0          0        0        0   
11422       0        0  ...        0        0          0        0        0   
11423       0        0  ...        0        0          0        0        0   
11424       0        0  ...        0        0          0        0        0   
11425       0        0  ...        0        0          0        0        0   

       ứng đáp  ứng đúng  ứng được  ứng đầy  ứng đủ  
0            0         0         0        0       0  
1            0         0         0        0       0  
2            0         0         0        0       0  
3            0         0         0        0       0  
4            0         0         0        0       0  
...        ...       ...       ...      ...     ...  
11421        0         0         0        0       0  
11422        0         0         0        0       0  
11423        0         0         0        0       0  
11424        0         0         0        0       0  
11425        0         0         0        0       0  

[11426 rows x 31384 columns]

In [26]:
word_frequency_sorted_bigram = word_frequency_table_bigram.sum(axis=0).sort_values()
# Nhưng từ xuất hiện ít nhất
word_frequency_sorted_bigram[:50]

1h30 mới            1
1h mới              1
1h có               1
10 năm              1
10 mới              1
10 mấy              1
10 luôn             1
10 kiến             1
10 fraction         1
10 bài              1
ứng nhưng           1
ổn bổ               1
ồn đến              1
ồn vang             1
100 là              1
100 cách            1
20 nhóm             1
10h30 nhưng         1
ứng kịp             1
10 50               1
ức chế              1
ủng hộ              1
ủa mà               1
10 trong            1
10 thì              1
10 sinh             1
10 phút             1
12 người            1
11h55 trước         1
11h30 nghỉ          1
11doubledot55 pm    1
20 25               1
2012 các            1
2011 một            1
2010 để             1
2010 hoặc           1
2010 gì             1
2009 và             1
2009 tới            1
2008 không          1
2008 hay            1
2008 2010           1
2007 hết            1
2005 để             1
2005 nên            1
19 20               1
15 đến              1
15 trong            1
30 sinh             1
30 giây             1
dtype: int64

In [27]:
# Những từ xuất hiện nhiều nhất
word_frequency_sorted_bigram[-20:]

tận tình       372
nội dung       374
lý thuyết      375
với sinh       394
thời gian      399
dạy rất        405
giáo viên      466
truyền đạt     507
tận tâm        563
thầy dạy       645
cho sinh       656
môn học        688
thực hành      877
kiến thức      904
giảng dạy      956
dễ hiểu       1004
bài tập       1057
giảng viên    1610
nhiệt tình    1848
sinh viên     2698
dtype: int64

**Kết luận**:
-  Với n = 2, cụm từ xuất hiện nhiều nhất là "sinh viên", những cụm từ xuất hiện ít nhất như: "1h30 mới", "2010 hoặc",...

### What are the limitations of this data processing flow ? How can we overcome those ?

You should comment your answer to problem 1 here with sufficient explanations, including your implementation and reasoning.

- Với cách thực hiện bằng phương pháp đếm số lần xuất hiện sẽ gặp những cụm từ thường xuyên xuất hiện nhưng vô nghĩa như: "cho sinh" xuất hiện 656 lần, "dạy rất" xuất hiện 405 lần. Ngoài ra, khi các từ trong cụm từ đứng riêng biệt thì sẽ không mang ý nghĩa. \
  -> ***Hướng giải quyết**: có thể sử dụng các bộ từ stopwords có sẵn để loại bỏ các cột chứa từ vô nghĩa*
- Trong trường hợp xuất hiện từ vựng nằm ngoại bộ từ vựng ban đầu, mô hình không có khả năng biểu diễn từ vựng mới trong vector embedding.

In [46]:
corpus = [
    'This is the first document.',
    'This document is the second document.',
]

count_model = CountVectorizer(ngram_range = (1, 1))
docs_vectorized = count_model.fit_transform(corpus).toarray()

print(docs_vectorized)

new_doc = ['This is the new document.']
new_doc_vectorized = count_model.transform(new_doc)
# Bộ dữ liệu ban đầu chỉ có 6 từ. Từ 'new' xuất hiện trong new_doc không được biểu diễn
new_doc_vectorized.toarray()


[[1 1 1 0 1 1]
 [2 0 1 1 1 1]]


array([[1, 0, 1, 0, 1, 1]])

- Khi tăng số lượng gram thì kích thước của ma trận biểu diễn một document tăng lên rất nhiều và tốn thời gian để xử lý. Ở ví dụ bên dưới, ta chọn số gram từ 1 -> 6



In [50]:
large_gram_model = CountVectorizer(ngram_range = (1, 4))
docs_vectorized = large_gram_model.fit_transform(corpus).toarray()

# Số chiều ban đầu của vector mô tả 1 document là 1x6. Trong trường hợp này là 1x26
print(docs_vectorized)


[[1 0 0 0 1 1 1 1 1 1 0 0 0 0 1 1 1 0 0 1 0 0 0 1 1 1]
 [2 1 1 1 0 0 1 1 0 0 1 1 1 1 1 0 0 1 1 1 1 1 1 0 0 0]]


## Stopwords

In [28]:
# Retrieve the stopword dictionary
!wget --no-check-certificate --content-disposition https://raw.githubusercontent.com/stopwords/vietnamese-stopwords/master/vietnamese-stopwords.txt

--2025-03-28 08:15:58--  https://raw.githubusercontent.com/stopwords/vietnamese-stopwords/master/vietnamese-stopwords.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20475 (20K) [text/plain]
Saving to: ‘vietnamese-stopwords.txt’

vietnamese-stopword 100%[===================>]  20.00K  --.-KB/s    in 0.002s  

2025-03-28 08:15:58 (11.2 MB/s) - ‘vietnamese-stopwords.txt’ saved [20475/20475]



In [29]:
# Observe stopwords list
vietnamese_stopword = open('vietnamese-stopwords.txt', 'r').read()
vietnamese_stopword = vietnamese_stopword.split('\n') # Separate lines by lines
print(f"#Number of stop words: {len(vietnamese_stopword)}")

#Number of stop words: 1942


In [30]:
# Stop words example
for sentence in vietnamese_stopword[:100]:
    print(sentence)

a lô
a ha
ai
ai ai
ai nấy
ai đó
alô
amen
anh
anh ấy
ba
ba ba
ba bản
ba cùng
ba họ
ba ngày
ba ngôi
ba tăng
bao giờ
bao lâu
bao nhiêu
bao nả
bay biến
biết
biết bao
biết bao nhiêu
biết chắc
biết chừng nào
biết mình
biết mấy
biết thế
biết trước
biết việc
biết đâu
biết đâu chừng
biết đâu đấy
biết được
buổi
buổi làm
buổi mới
buổi ngày
buổi sớm
bà
bà ấy
bài
bài bác
bài bỏ
bài cái
bác
bán
bán cấp
bán dạ
bán thế
bây bẩy
bây chừ
bây giờ
bây nhiêu
bèn
béng
bên
bên bị
bên có
bên cạnh
bông
bước
bước khỏi
bước tới
bước đi
bạn
bản
bản bộ
bản riêng
bản thân
bản ý
bất chợt
bất cứ
bất giác
bất kì
bất kể
bất kỳ
bất luận
bất ngờ
bất nhược
bất quá
bất quá chỉ
bất thình lình
bất tử
bất đồ
bấy
bấy chầy
bấy chừ
bấy giờ
bấy lâu
bấy lâu nay
bấy nay
bấy nhiêu
bập bà bập bõm
bập bõm
bắt đầu
bắt đầu từ


In [31]:
len(vietnamese_stopword)

1942

In [32]:
len(set(vietnamese_stopword) & set(word_frequency_table_bigram.columns.to_list()))

500

## Term frequency - Invert document frequency (TF-IDF)


### TF
Term frequency (TF) is the number of times a given term appears in document

$$
tf(t) = f(t,d)\times\frac{1}{T}
$$
whereas, $f(t,d)$ is the frequency of the word $t$ in the document $d$, $T$ is the number of all words in that document.

In [51]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer

corpus = [
    'This is the first document.',
    'This document is the second document.',
    'And this is the third one.',
    'Is this the first document?',
]

# Declare TF vectorize
tf_vectorizer = TfidfVectorizer(ngram_range=(1, 1),
                                use_idf=False, # only using TF
                                norm='l1')

tf_vectorizer.fit_transform(corpus)

tf_vectorized = tf_vectorizer.transform(corpus)

tf_output = tf_vectorized[0]

# Build TF table
words_tf_idf = pd.DataFrame(tf_output.T.todense(), index=tf_vectorizer.get_feature_names_out(), columns=['tf'])
words_tf_idf

tf
and       0.0
document  0.2
first     0.2
is        0.2
one       0.0
second    0.0
the       0.2
third     0.0
this      0.2

### IDF

Inverse Document Frequency, or abbreviated as IDF, measures how important a term is. While computing TF, all terms are considered equally important. However it is known that certain terms, such as "is", "of", and "that", may appear a lot of times but have little importance. Thus we need to weigh down the frequent terms while scale up the rare ones.

$$
idf(t) = \log\left(\frac{\text{#documents in the document set}}{\text{#documents with term}}\right) + 1
$$

In [52]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer

corpus = [
    'This is the first document.',
    'This document is the second document.',
    'And this is the third one.',
    'Is this the first document?',
]

# Configure settings for IDF vectorize
tf_idf_vectorizer = TfidfVectorizer(ngram_range=(1, 1),
                                    smooth_idf=False,
                                    use_idf=True,
                                    norm=None)

tf_idf_vectorizer.fit_transform(corpus)

# Retrieve only idf information
idf_vectorizer = tf_idf_vectorizer.idf_

# Join idf values into the previous dataframe
words_tf_idf['idf'] = idf_vectorizer

# Show dataframe with ascending values of idf
words_tf_idf.sort_values(by=['idf'])

tf       idf
is        0.2  1.000000
the       0.2  1.000000
this      0.2  1.000000
document  0.2  1.287682
first     0.2  1.693147
and       0.0  2.386294
second    0.0  2.386294
one       0.0  2.386294
third     0.0  2.386294

### TF-IDF

Technically saying, TF-IDF is a score which is applied to every word in every document in our dataset. And for every word, the TF-IDF value increases with every appearance of the word in a document, but is gradually decreased with every appearance in other documents

$$
\text{tf-idf}= tf(t, d) \times idf(t)
$$

In [53]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer

corpus = [
    'This is the first document.',
    'This document is the second document.',
    'And this is the third one.',
    'Is this the first document?',
]

tf_idf_vectorizer = TfidfVectorizer(ngram_range=(1, 1),
                                    smooth_idf=True,
                                    use_idf=True,
                                    norm='l1')

tf_idf_vectorizer.fit_transform(corpus)

tf_idf_vectorized = tf_idf_vectorizer.transform(corpus)

tf_idf_output = tf_idf_vectorized[0]
words_tf_idf['tf-idf'] = tf_idf_output.T.todense()

words_tf_idf.sort_values(by=['tf-idf'])

tf       idf    tf-idf
and       0.0  2.386294  0.000000
third     0.0  2.386294  0.000000
second    0.0  2.386294  0.000000
one       0.0  2.386294  0.000000
is        0.2  1.000000  0.174399
the       0.2  1.000000  0.174399
this      0.2  1.000000  0.174399
document  0.2  1.287682  0.213315
first     0.2  1.693147  0.263487

### Problem set 2
Based on the problem 1 and the instruction on TF, IDF, TF-IDF:
- (2a) Build the tf-idf table for the UIT-VSFC dataset with $n$-gram = 1 and $n$-gram = 2.
- (2b) Change a few hyperparameters in the `TfidfVectorizer` function (`smooth_idf`, `sublinear_tf` and `norm`) from problem 2a (*you could browse from this [link](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html) to discover which are the correct paramters to parse*). Explain the results differences collected after modifying hyperparameters.
- (2c) Which words has the lowest and the highest tf-idf values ? Do they differ from $n$-grams results ?
- (2d) Which limitations from $n$-grams that TF-IDF overcame ?

**(2a) Build the tf-idf table for the UIT-VSFC dataset with  n -gram = 1 and  n -gram = 2.**

In [54]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer

def create_tf_idf_table(dataset: list,
                        n: int = 1,
                        smooth_idf: bool = True,
                        sublinear_tf: bool = True,
                        norm: str = 'l2') -> pd.DataFrame:

    """
    Function to create a term frequency-inverse document frequency for a dataset

    Args:
    dataset -- A subset (i.e., train/valid/test) in UIT-VSFC dataset
    n (int) -- Number of grams that we parse into this function
    smooth_idf (bool) -- Smoothing parameter for IDF
    sublinear_tf (bool) --

    Returns:
    pd.DataFrame -- A dataframe contains all words after conducting n-grams and their respective frequencies

    """
    tf_idf_vectorizer = TfidfVectorizer(ngram_range=(n, n),
                                        smooth_idf=smooth_idf,
                                        sublinear_tf=sublinear_tf,
                                        use_idf=True,
                                        norm=norm)
    tf_idf_output = tf_idf_vectorizer.fit_transform(dataset)
    tf_idf_output_df = pd.DataFrame(tf_idf_output.toarray(), columns=tf_idf_vectorizer.get_feature_names_out())
    return tf_idf_output_df



In [55]:
tf_idf_table_unigram = create_tf_idf_table(dataset=list_all_sentence,
                                            n=1,
                                            smooth_idf=True,
                                            sublinear_tf=False,
                                            norm='l2')
tf_idf_table_unigram

10  100  10h  10h30   11  11doubledot55  11h30  11h55   12  \
0      0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
1      0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
2      0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
3      0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
4      0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
...    ...  ...  ...    ...  ...            ...    ...    ...  ...   
11421  0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
11422  0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
11423  0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
11424  0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
11425  0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   

       12doubledot00  ...   ấy   ẩn   ắt   ốc   ồn   ổn   ủa  ủng   ức  ứng  
0                0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1                0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2                0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3                0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4                0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
...              ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  
11421            0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
11422            0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
11423            0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
11424            0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
11425            0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[11426 rows x 2459 columns]

In [56]:
tf_idf_table_bigram = create_tf_idf_table(dataset=list_all_sentence,
                                            n=2,
                                            smooth_idf=True,
                                            sublinear_tf=False,
                                            norm='l2')
tf_idf_table_bigram

10 50  10 bài  10 fraction  10 kiến  10 luôn  10 mấy  10 mới  10 người  \
0        0.0     0.0          0.0      0.0      0.0     0.0     0.0       0.0   
1        0.0     0.0          0.0      0.0      0.0     0.0     0.0       0.0   
2        0.0     0.0          0.0      0.0      0.0     0.0     0.0       0.0   
3        0.0     0.0          0.0      0.0      0.0     0.0     0.0       0.0   
4        0.0     0.0          0.0      0.0      0.0     0.0     0.0       0.0   
...      ...     ...          ...      ...      ...     ...     ...       ...   
11421    0.0     0.0          0.0      0.0      0.0     0.0     0.0       0.0   
11422    0.0     0.0          0.0      0.0      0.0     0.0     0.0       0.0   
11423    0.0     0.0          0.0      0.0      0.0     0.0     0.0       0.0   
11424    0.0     0.0          0.0      0.0      0.0     0.0     0.0       0.0   
11425    0.0     0.0          0.0      0.0      0.0     0.0     0.0       0.0   

       10 năm  10 phút  ...  ứng kịp  ứng nhu  ứng nhưng  ứng tốt  ứng yêu  \
0         0.0      0.0  ...      0.0      0.0        0.0      0.0      0.0   
1         0.0      0.0  ...      0.0      0.0        0.0      0.0      0.0   
2         0.0      0.0  ...      0.0      0.0        0.0      0.0      0.0   
3         0.0      0.0  ...      0.0      0.0        0.0      0.0      0.0   
4         0.0      0.0  ...      0.0      0.0        0.0      0.0      0.0   
...       ...      ...  ...      ...      ...        ...      ...      ...   
11421     0.0      0.0  ...      0.0      0.0        0.0      0.0      0.0   
11422     0.0      0.0  ...      0.0      0.0        0.0      0.0      0.0   
11423     0.0      0.0  ...      0.0      0.0        0.0      0.0      0.0   
11424     0.0      0.0  ...      0.0      0.0        0.0      0.0      0.0   
11425     0.0      0.0  ...      0.0      0.0        0.0      0.0      0.0   

       ứng đáp  ứng đúng  ứng được  ứng đầy  ứng đủ  
0          0.0       0.0       0.0      0.0     0.0  
1          0.0       0.0       0.0      0.0     0.0  
2          0.0       0.0       0.0      0.0     0.0  
3          0.0       0.0       0.0      0.0     0.0  
4          0.0       0.0       0.0      0.0     0.0  
...        ...       ...       ...      ...     ...  
11421      0.0       0.0       0.0      0.0     0.0  
11422      0.0       0.0       0.0      0.0     0.0  
11423      0.0       0.0       0.0      0.0     0.0  
11424      0.0       0.0       0.0      0.0     0.0  
11425      0.0       0.0       0.0      0.0     0.0  

[11426 rows x 31384 columns]

**(2b) Explain the results differences collected after modifying hyperparameters.**

- smooth_idf=True: Theo tài liệu của thư viện scikitlearn, công thức mặc định là idf(t) =  log [ (1 + n) / (1 + df(t)) ] + 1. Điều này đảm bảo tránh trường hợp chia cho 0. Nếu tham số này nhận giá trị False, giá trị idf được tính theo công thức truyền thống là idf(t) = log(n/df(t)) + 1
- sublinear_tf=False: Theo tài liệu của thư việc scikitlearn, mặc định tf sẽ được tính tần số như bình thường. Khi giá trị này sẽ được tính bởi log.
- norm: Chuẩn hóa vector document theo chuẩn L2, L1, hoặc Không chuẩn hóa

**(2c) Which words has the lowest and the highest tf-idf values ? Do they differ from  n -grams results ?**

- Với n = 1, một số từ có giá trị tf-idf cao nhất như: thân, nope, được, like, nghiệm, hay, vui, tốt, tệ,... Các tự có giá trị tf-idf thấp như: 2048, 1dot5, 1983,...

In [57]:
max_ind = tf_idf_table_unigram.max().sort_values(ascending=False)
print(max_ind[:50])

dễ            1.000000
tệ            1.000000
everything    1.000000
giỏi          1.000000
hết           1.000000
vui           1.000000
hay           1.000000
hi            1.000000
hiền          1.000000
tốt           1.000000
chán          1.000000
nghiêm        1.000000
tuyệt         1.000000
listening     1.000000
like          1.000000
xinh          1.000000
không         1.000000
nope          1.000000
thân          1.000000
được          1.000000
điểm          1.000000
dở            0.967450
love          0.967144
chào          0.967144
ổn            0.953973
thờ           0.947277
nản           0.932821
chậm          0.928421
tánh          0.922884
thày          0.919391
ok            0.918819
nhóm          0.907757
nhiêt         0.907344
co            0.906940
chơi          0.904809
nhập          0.903042
nhộn          0.901618
chịu          0.898115
nhanh         0.895971
day           0.894392
colonhihi     0.892059
vỉên          0.891448
diễu          0.886464
dám        

In [58]:
min_ind = tf_idf_table_unigram.min().sort_values(ascending=True)
print(min_ind[-30:])

đồng     0.0
đồm      0.0
đồ       0.0
2048     0.0
2015     0.0
2013     0.0
2012     0.0
2011     0.0
2010     0.0
2009     0.0
2008     0.0
2007     0.0
2005     0.0
2003     0.0
20       0.0
1h30     0.0
1h       0.0
1dot5    0.0
1983     0.0
ấy       0.0
ẩn       0.0
ắt       0.0
ốc       0.0
ồn       0.0
ổn       0.0
ủa       0.0
ủng      0.0
ức       0.0
ứng      0.0
10       0.0
dtype: float64


Với n = 2, một số từ có giá trị tf-idf cao nhất như: hết mình, bài giảng, tâm huyết, thảo mai, dạy vui,... Các tự có giá trị tf-idf thấp như: ứng đủ, 10 50, 10 bài, 10 fraction,...

In [59]:
max_ind = tf_idf_table_bigram.max().sort_values(ascending=False)
print(max_ind[:50])

quá tốt          1.0
quá nhanh        1.0
không rõ         1.0
ngắn gọn         1.0
giảng bài        1.0
còn nhanh        1.0
hi hi            1.0
về muộn          1.0
bình thường      1.0
mọi thứ          1.0
dễ chịu          1.0
chưa ổn          1.0
dạy chán         1.0
hiền hậu         1.0
thân ái          1.0
thân thiện       1.0
rất nản          1.0
không có         1.0
đầy đủ           1.0
giảng dạy        1.0
vui nhộn         1.0
chuyên nghiệp    1.0
thầy hiền        1.0
dễ hiểu          1.0
cám ơn           1.0
khổ lắm          1.0
không biết       1.0
sơ sài           1.0
về trễ           1.0
hài hước         1.0
tâm huyết        1.0
dạy chậm         1.0
hơi căng         1.0
hơi khó          1.0
giảng hay        1.0
giáo viên        1.0
dạy dở           1.0
kể chuyện        1.0
rất tốt          1.0
hòa đồng         1.0
hết mình         1.0
hấp dẫn          1.0
sinh động        1.0
cảm ơn           1.0
quan tâm         1.0
áp dụng          1.0
kiến thức        1.0
khó nghe     

In [60]:
min_ind = tf_idf_table_bigram.min().sort_values(ascending=False)
print(min_ind[:30])

ứng đủ         0.0
10 50          0.0
10 bài         0.0
10 fraction    0.0
10 kiến        0.0
10 luôn        0.0
10 mấy         0.0
10 mới         0.0
ổn thỏa        0.0
ổn thầy        0.0
ổn không       0.0
ổn có          0.0
ổn bổ          0.0
ồn đến         0.0
ồn ào          0.0
ồn vang        0.0
ồn sẽ          0.0
ồn quá         0.0
ồn nên         0.0
ồn mệt         0.0
ồn mong        0.0
ồn do          0.0
ồn cô          0.0
ốc nên         0.0
15 đến         0.0
15 trong       0.0
15 phút        0.0
15 người       0.0
15 câu         0.0
15 25          0.0
dtype: float64


**(2d) Which limitations from  n -grams that TF-IDF overcame ?**

- Đối với phương pháp TF-IDF, không chỉ dựa vào tần số xuất hiện của từ mà còn dựa vào IDF. Đều này có thể giúp TF-IDF không gán trọng số cao cho các từ thường xuyên xuất hiện nhưng ít mang ý nghĩa (stopwords):
- Ngoài ra, TF-IDF còn gán cho trọng số cao cho những từ xuất hiện nhiều trong một document cụ thể như ít xuất hiện đồng thời trong nhiều document.